bagel/derivatives_utils.py

```diff
@@ -0,0 +1,89 @@
+}
+
+
+def check_pipelines_are_recognized(pipelines: Iterable[str]):
+    """Check that all pipelines in the processing status file are supported by Nipoppy."""
+    unrecognized_pipelines = list(
+        set(pipelines).difference(mappings.get_pipeline_uris())
+    )
+    if len(unrecognized_pipelines) > 0:
+        raise LookupError(
+            f"The processing status file contains unrecognized pipelines in the column '{PROC_STATUS_COLS['pipeline_name']}': "
+            f"{unrecognized_pipelines}. "
+            f"Allowed pipeline names are the following pipelines supported natively in Nipoppy (https://github.com/nipoppy/pipeline-catalog): \n"
+            f"{mappings.get_pipeline_uris()}"
+        )
+
+
+def check_pipeline_versions_are_recognized(
+    pipeline: str, versions: Iterable[str]
+):
+    """
+    Check that all pipeline versions in the processing status file are supported by Nipoppy.
+    Assumes that the input pipeline name is recognized.
+    """
+    unrecognized_versions = list(
+        set(versions).difference(mappings.get_pipeline_versions()[pipeline])
+    )
+    if len(unrecognized_versions) > 0:
+        raise LookupError(
+            f"The processing status file contains unrecognized {pipeline} versions in the column '{PROC_STATUS_COLS['pipeline_version']}': {unrecognized_versions}. "
+            f"Allowed {pipeline} versions are the following versions supported natively in Nipoppy (https://github.com/nipoppy/pipeline-catalog): \n"
+            f"{mappings.get_pipeline_versions()[pipeline]}"
+        )
+
+
+def get_subject_imaging_sessions(
+    jsonld_subject: models.Subject,
+) -> dict:
+    """
+    Return a dictionary of imaging sessions for a given subject from JSONLD data,
+    where the keys are the session labels and values are the session objects.
+    """
+    jsonld_sub_sessions_dict = {}
+    for jsonld_sub_ses in getattr(jsonld_subject, "hasSession"):
+        if jsonld_sub_ses.schemaKey == "ImagingSession":
+            jsonld_sub_sessions_dict[jsonld_sub_ses.hasLabel] = jsonld_sub_ses
+
+    return jsonld_sub_sessions_dict
+
+
+def create_completed_pipelines(session_proc_df: pd.DataFrame) -> list:
+    """
+    Create a list of CompletedPipeline objects for a single subject-session based on the completion status
+    info of pipelines for that session from the processing status dataframe.
+    """
+    completed_pipelines = []
+    for (pipeline, version), session_pipe_df in session_proc_df.groupby(
+        [
+            PROC_STATUS_COLS["pipeline_name"],
+            PROC_STATUS_COLS["pipeline_version"],
+        ]
+    ):
+        # Check that all pipeline steps have succeeded
+        if (
+            session_pipe_df[PROC_STATUS_COLS["status"]].str.lower()
+            == "success"
+        ).all():
+            completed_pipeline = models.CompletedPipeline(
+                hasPipelineName=models.Pipeline(
+                    identifier=mappings.get_pipeline_uris()[pipeline]
```

Comment:
Ok. Could you remind me again why we don't want to run + store the output of `get_pipeline_uris()` and `get_pipeline_versions()` in mappings.py in variables like 
```
KNOWN_PIPELINES = get_pipeline_uris()
KNOWN_VERSIONS = get_pipeline_uris()
```
and then just import these variables into other modules?
Since these funcs just rely on the submodule, the catalog files should essentially always be available when a user runs the CLI, no?

Reply:
Also, in this specific function, it's only used once, so I'm not sure it's worth storing it in a separate local variable.

Reply:
> Ok. Could you remind me again why we don't want to run + store the output of `get_pipeline_uris()` and `get_pipeline_versions()` in mappings.py in variables like
> 
> ```
> KNOWN_PIPELINES = get_pipeline_uris()
> KNOWN_VERSIONS = get_pipeline_uris()
> ```
> 
> and then just import these variables into other modules?

I guess that could be justified, yes. Generally you want to avoid side-effects when importing a module so it can be used flexibly in contexts that may e.g. not require the config. So it's a good pattern to avoid top-level code execution in modules except for constants and such. But that's more true for packages. CLI is an app, and the config is needed for the app to run. Essentially what you're saying is: if you didn't setup the pipeline-catalog submodule and you run some part of the code/app (e.g. during testing) that imports this module for whatever reason, we're OK with that crashing. And that's probably OK.